netmiko/cisco/cisco_xr.py

```diff
@@ -105,36 +114,74 @@ def commit(
-        new_data = self._send_command_str(
-            command_string,
-            expect_string=r"(#|onfirm)",
-            strip_prompt=False,
-            strip_command=False,
-            read_timeout=read_timeout,
-        )
-        if "onfirm" in new_data:
-            output += new_data
+        if replace:
+            new_data = self._send_command_str(
+                command_string,
+                expect_string=r"This commit will replace or remove the entire running configuration",
+                strip_prompt=False,
+                strip_command=False,
+                read_timeout=read_timeout,
+            )
+            if "This commit will replace or remove the entire running configuration" in new_data:
+                output += new_data
+                new_data = self._send_command_str(
+                    "yes",
+                    expect_string=r"#",
+                    strip_prompt=False,
+                    strip_command=False,
+                    read_timeout=read_timeout,
+                )
+        else:
+            # IOS-XR might do this:
+            # This could be a few minutes if your config is large. Confirm? [y/n][confirm]
             new_data = self._send_command_str(
-                "y",
-                expect_string=r"#",
+                command_string,
+                expect_string=r"(#|onfirm)",
                 strip_prompt=False,
                 strip_command=False,
                 read_timeout=read_timeout,
             )
+            if "onfirm" in new_data:
+                output += new_data
+                new_data = self._send_command_str(
+                    "y",
+                    expect_string=r"#",
+                    strip_prompt=False,
+                    strip_command=False,
+                    read_timeout=read_timeout,
+                )
         output += new_data
         if error_marker in output:
             raise ValueError(f"Commit failed with the following errors:\n\n{output}")
         if alt_error_marker in output:
+            if commit_error_dialog_dict is not None and alt_error_marker in commit_error_dialog_dict:
+                marker_value = commit_error_dialog_dict[alt_error_marker]
+                output += new_data
+                new_data += self._send_command_str(
+                    marker_value,
+                    expect_string=r"\)#$",
+                    strip_prompt=False,
+                    strip_command=False,
+                    read_timeout=read_timeout,
+                )
+                output += new_data
+            else:
             # Other commits occurred, don't proceed with commit
-            output += self._send_command_timing_str(
-                "no", strip_prompt=False, strip_command=False
-            )
-            raise ValueError(f"Commit failed with the following errors:\n\n{output}")
+                output += self._send_command_timing_str(
```

Comment:
use _send_command_str here as well, timing methods should be avoided as we know what prmpt to expect

Reply:
> At the end @akshansh237 and @shreyashah , I think all the debugging from my oct-dec,2022 didnt translate here or between you guys it got lost. I am okay with this PR but the netmiko is slowing us down and whenever and wherever it is necessary we must get to the failure first. So, I am okay if this code goes in , but immediate next one should have a while loop or wait must and must be checked to see if session is alive. Error pattern should say :
> 
> 1. rtr session not alive.
> 2. Rtr session alive but no prompt found. This should execute some traces collection from new channel.
> 3. any other error.




Both 1 and 2 are added in the code. You can check send_command(),  read_until_pattern() functions in base_connection.py